### ISSUE TITLE
Convert stills_process to use ImageSequence

### ISSUE THREAD

**Issue by dwpaley** [dials/dials#2377]
As per comment here: https://github.com/cctbx/dxtbx/pull/578#issuecomment-1317371352

For now, just want to give this an issue number.

**Comment by dwmoreau** [comment]
I've created branches on dials & dxtbx, `stills_process_imagesequence`, that demonstrate a prototype to convert stills_process to use ImageSequence. I don't expect this branch to be merged into main; its purpose is to surface the concrete changes, new objects, and sharp edges a production effort will hit. On the other hand, dials.stills_process does run on these new branches with experiment files properly formatted as ImageSequence.

I do want to acknowledge previous work done on this conversion: https://github.com/dials/dials/tree/development_dsp_sequence_test/src/dials

The rest of this document is mostly AI generated, as is most of the code on these branches. I will commit these branches when NERSC is back up from maintenance.

## What the branch does

`dials.stills_process` currently converts every imported image into a plain `ImageSet` in `do_import()`. This branch keeps the data as an `ImageSequence` (or the new `XFELImageSequence`) for the entire pipeline. Two consequences follow:

- **Smaller output.** Experiments from one source file can share a single imageset/beam/detector object, so the `.expt` file stores one copy instead of one per frame.
- **Potentially faster runtime.** Per-frame caching of the spotfinder and indexer becomes possible, because the cached objects now see a stable imageset identity across frames.
- **Downstream performance improvements.** Potentially, programs like cctbx.xfel.merge and dials.image_viewer will get a boost in performance due to an easier parsing of the experiment file.

The rest of this document is the concrete list of what changed and why.

## 1. The pivot: `do_import()` no longer converts to `ImageSet`

**Before:** `do_import()` called `ExperimentListFactory.from_filenames()` and then `ImageSetFactory.imageset_from_anyset(experiment.imageset)`. That conversion rebuilt the image data and handed back fresh, per-frame beam/detector model **copies**.

**After:** for multi-image formats it calls `format_class.get_imageset(as_sequence=True)` directly, yielding an `ImageSequence` whose beam/detector are shared-identity Python objects. The per-frame loop slices `imageset[i:i+1]` and keeps each slice an `ImageSequence`.

**Why this was the blocker.** The old conversion destroyed *object identity*. The JSON serializer deduplicates models by identity — same Python object written once. The spotfinder mask cache (section 5) also keys on the imageset object. With per-frame copies, neither could work. Removing the conversion is the root change; everything else follows from it.

## 2. New dxtbx objects

**`XFELImageSequence`** (Python, subclass of `ImageSequence`, in `imageset.py`). An XFEL run has a different wavelength per pulse, but a base `ImageSequence` carries a single beam. `XFELImageSequence` stores a per-frame `wavelengths` list plus an optional `XFELBeam`. `get_beam(i)` returns a monochromatic `Beam` for frame `i`; `get_beam()` with no index returns the `XFELBeam`. Slicing and pickling are overridden to carry the wavelength list through.

**`XFELBeam`** (C++, extends `Beam`, in `model/beam.h` / `boost_python/beam.cc`). Holds an array of per-frame wavelengths. `get_wavelength()`, `set_wavelength()`, and `get_s0()` deliberately **throw** — there is no single wavelength. Serializes with `__id__: "xfel"`; `BeamFactory.from_dict` routes that id to it. It is constructed via `BeamFactory.make_xfel_beam()` (new). A Python-side helper `get_beam_at_frame(fi)` (injected onto `XFELBeam` in `model/__init__.py`) returns the monochromatic `Beam` for a given frame — this is the access pattern a downstream `.expt` consumer must use **after a JSON round-trip**, since the on-disk beam is the `XFELBeam`, not a per-frame `Beam`. Output `.expt` files get **one `XFELBeam` per source file** instead of N monochromatic beams. This is the most consequential item — it is a genuine new physical model that every downstream `.expt` consumer will see, and the one most worth group discussion.

**`FormatXFEL` mixin + `FormatNXmxXFEL` / `FormatXTCXFEL`** (in `format/`). The mixin's `get_imageset()` lets the parent format build the raw `ImageSet`, reads per-frame wavelengths via a subclass `get_wavelengths()`, builds the `XFELBeam`, and returns an `XFELImageSequence`. For NXmx, `get_wavelengths()` goes through `CachedWavelengthBeamFactory.make_xfel_beam()` (new, in `dxtbx/nexus/__init__.py`), which performs the actual 1-D `incident_wavelength` HDF5 read. `FormatNXmxXFEL.understand()` distinguishes XFEL data (1-D `incident_wavelength` array) from synchrotron data (scalar); `FormatXTCXFEL` delegates to `FormatXTC` since XTC streams are always XFEL stills.

## 3. Cross-cutting change: how a "still" is detected

Once stills stay as `ImageSequence`, the old test `isinstance(imageset, ImageSequence)` no longer separates stills from rotations — **both are now `ImageSequence`**. Every such check had to move to the scan:

> a still is `scan is None or scan.is_still()` (zero-oscillation).

Touched: `Indexer.from_parameters` (stills/sequence routing), `indexer._xyzcal_mm_to_px`, `reflection_predictor` (static vs scan-varying predictor dispatch), `integrator._determine_max_memory_needed`, and `SpotFinderFactory` (via a new `is_stills` argument). It also let an entire block be **deleted** from `Indexer.from_parameters` — the code that used to convert `ImageSequence`→`ImageSet` and null scan/goniometer is now unnecessary.

This is the widest blast radius in the branch and the change most in need of a pipeline-wide decision: is scan-based detection the contract we want everywhere?

## 4. Output: shared models per source file

`_rebuild_shared_imageset_output()` (new, in `stills_process.py`) runs before writing. It groups experiments by source file and rebuilds them so every experiment from one file points at **one** `ImageSequence` / beam / detector object, and multi-lattice experiments from the same frame share **one** `Scan` object. This is the change that actually shrinks the file. It is a no-op for rotation data. A helper `_frame_index()` (new) recovers a frame's 0-based index, handling both representations — the per-frame imageset from `do_import` and the full shared imageset after a JSON round-trip.

`combine_all_ranks` (new phil option) merges the per-rank (MPI) or per-worker (multiprocessing) composite files into a single combined file with shared models. The multiprocessing path is handled by `_combine_multiprocessing_outputs()` (new). As part of this, `extend_with_bookkeeping` — which remaps reflection `id` and `experiment_identifiers` when merging — was lifted out of the MPI `finalize` branch into a module-level function so both the MPI and multiprocessing paths reuse it.

## 5. Performance caching (unblocked by section 1)

**Spotfinder** (`finder.py` + `stills_process.py`). This is a reimplementation of https://github.com/dials/dials/pull/3157. `ExtractPixelsFromImage` splits its mask into a *static* component (lookup mask + untrusted rectangles, frame-independent) and a *dynamic* per-frame component (`imageset.get_mask(index)`). A new `update_imageset()` swaps in the next frame's data without rebuilding. `ExtractSpots` caches the `ExtractPixelsFromImage` object and reuses it across stills frames. The `Processor` itself also caches the configured `SpotFinderFactory` in `self.spot_finder_factory` (built once by the new `get_spot_finder_factory()`), so the factory is not reconstructed per frame.

**Indexer** (`stills_process.py`). This is a reimplementation of https://github.com/dials/dials/pull/3158. `Processor` caches the `Indexer` object (`self.idxr`, `self.idxr_method_list`, `self.idxr_known_crystal_models`); a new `update_indexer()` resets per-frame state instead of reconstructing. `copy.deepcopy` of the params object per image was replaced with direct mutate-and-restore of two fields: `refinement.parameterisation.scan_varying` (forced `False`) and `basis_vector_combinations.max_refine`. Note the latter is now pinned to `5` for stills indexing — a behavioral change, not just a refactor — and restored afterward.

Both caches depend on section 1 — they only work because the imageset identity is now stable across frames.

## 6. Surprises / sharp edges

- The **C++ `ImageSequence` requires a non-null scan**, but stills have none. Two separate code paths work around this, and they should not be conflated: (a) for XFEL data, `FormatXFEL.get_imageset()` builds a **real** zero-oscillation `Scan((1, n), (0.0, 0.0))` directly when constructing the `XFELImageSequence` — there is no error to catch; (b) for non-XFEL multi-image formats, `format_class.get_imageset(as_sequence=True)` raises a `RuntimeError` (its `get_scan()` returns `None`), so `do_import` catches that, loads the data as an `ImageSet`, and wraps it in an `ImageSequence` with a **synthetic** zero-oscillation scan.
- **`load_models()` re-reads the synthetic still scan** back onto the experiment, so the scan/goniometer must be re-nulled after *every* `load_models()` call (both `do_import` and `do_work`).
- A **per-frame slice carries the absolute frame range** (e.g. `(3913, 3913)`), which conflicts with the integrator's expectation of 0-based z coordinates. The slice is rebuilt with a fresh `(1, 1)` scan.
- `XFELImageSequence.get_beam()` with no index returns an `XFELBeam`, which broke `get_s0()` in the spotfinder — callers must use `get_beam(i)` for a monochromatic beam.
- `easy_mp` tried to **pickle C++ threshold/indexer objects** in the result queue; `do_work` now returns `None` when `finalize=True` since output is already on disk. This is the fix in https://github.com/dials/dials/pull/3156.
- `flex.double` is not JSON-serializable; `XFELBeam.to_dict` converts to a Python list.

## 7. Next steps:
- Incorporate feedback from the broader DIALS group on high-level decisions. Obviously everything in this document is up for debate and scrutiny.
- Create an ImageSet -> ImageSequence converter for legacy experiment files
- Work in compatibility for all downstream applications
- Get all tests to pass.
- Create new tests for new objects and code.

**Comment by dwmoreau** [comment]
@dagewa & @toastisme gave feedback in a Slack message. They point out that the creation of XFELImageSequence and XFELBeam are problematic. The way I implemented XFELBeam forces the Beam to have knowledge of the images, their ordering. The beam should not need to know about the order of the frames. @toastisme solved this problem in a different way when working on the TOF/Laue case: [https://github.com/cctbx/dxtbx/pull/620](url) & [https://github.com/cctbx/dxtbx/pull/621](url). In the ToF-Laue case, every image has a different ToF/wavelength. @toastisme uses the scan to encode information. I am updating my prototype code to take an analogous approach. A new XFELBeam, similar to the PolychromaticBeam is created that just guards against certain methods. The beam itself is absent of wavelength, which is encoded in the scan.

A beam without wavelength is a useful concept for streaming from LCLS. Following the SIMPLON API, the LCLStreamer will send out a start message, followed by a series of image messages, then an end message. In the LCLStreamer case, the start message does not contain a beam wavelength. A wavelength-less beam can now be created from this start message.
